ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastHashTableLoader.java

```diff
@@ -106,9 +107,21 @@ public void init(ExecMapperContext context, MapredContext mrContext,
   }
 
   private void initHTLoadingService(long estKeyCount) {
-    // Avoid many small HTs that will rehash multiple times causing GCs
-    this.numLoadThreads = (estKeyCount < VectorMapJoinFastHashTable.FIRST_SIZE_UP) ? 1 :
-        HiveConf.getIntVar(hconf, HiveConf.ConfVars.HIVEMAPJOINPARALELHASHTABLETHREADS);
+    if (estKeyCount < VectorMapJoinFastHashTable.FIRST_SIZE_UP) {
+      // Avoid many small HTs that will rehash multiple times causing GCs
+      this.numLoadThreads = 1;
+    } else {
+      int initialValue = HiveConf.getIntVar(hconf, HiveConf.ConfVars.HIVEMAPJOINPARALELHASHTABLETHREADS);
+      Preconditions.checkArgument(initialValue > 0, "The number of HT-loading-threads should be positive.");
+
+      int adjustedValue = Integer.highestOneBit(initialValue);
+      if (initialValue != adjustedValue) {
+        LOG.info("Adjust the number of HT-loading-threads to {}. (Previous value: {})",
```

Comment:
what's the use of that, should we auto-tune?
````
 this.numLoadThreads=adjustedValue?
````

Reply:
Yes, I intended auto-tuning, and the code is wrong. Thanks for pointing it out.